### PR TITLE
Timescale Output Plugin for telegraf

### PR DESCRIPTION
**Pull request by blagojts**
Add a new tutorial for the Output Plugin we developed 
for Telegraf to write data into PostgreSQL or TimescaleDB.

### CODE REVIEW

File: tutorials/telegraf-output-plugin.md [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+And the output should look something like: 
+
+```
+> cpu,cpu=cpu0,host=local usage_guest=0,usage_idle=78.431372,usage_iowait=0,usage_irq=0,usage_softirq=0,usage_steal=0,usage_system=11.764705,usage_user=9.803921 1558613882000000000
+> cpu,cpu=cpu1,host=local usage_guest=0,usage_idle=92.156862,usage_iowait=0,usage_irq=0,usage_softirq=0,usage_steal=0,usage_system=3.921568,usage_user=3.921568 1558613882000000000
+> cpu,cpu=cpu-total,host=local usage_guest=0,usage_idle=87.623762,usage_iowait=0,usage_irq=0,usage_softirq=0,usage_steal=0,usage_system=6.435643,usage_user=5.940594 1558613882000000000
+```
+
+A line is outputted for each core of the CPU and the total. Values are presented in `key=value` pairs with the timestamp last in the row. 
+Writing to STDOUT doesn't distinguish between *tags*, which are indexed fields (cpu, host) and value *fields* (usage_quest, usage_user ...).
```
Comment (svenklemm) [inline tutorials/telegraf-output-plugin.md]: you can distinguish tags from fields in that output, there is a space between tags and fields

File: tutorials/telegraf-output-plugin.md [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+
+## Installation [](telegraf-installation)
+
+### Before we start
+
+Before we start, you will need [TimescaleDB installed][getting-started] and a means to connect to it.
+
+### Setting up Telegraf
+
+Telegraf is written in Go, so only one standalone binary is required to run it. But also because of this all the code for the different plugins must be part of that binary. We have an unofficial build of Telegraf version 1.10.4 with our plugin added. You can download from the following links:
```
Comment (svenklemm) [inline tutorials/telegraf-output-plugin.md]: First 2 sentences sound a bit weird, also i think go added support for plugins but telegraf does not use that infrastructure yet

File: tutorials/telegraf-output-plugin.md [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+
+### Setting up Telegraf
+
+Telegraf is written in Go, so only one standalone binary is required to run it. But also because of this all the code for the different plugins must be part of that binary. We have an unofficial build of Telegraf version 1.10.4 with our plugin added. You can download from the following links: 
+
+* Linux amd64: [deb][deb-build] [rpm][rpm-build] [binary][linux-bin-build]
+* Windows amd64 [binary/exe][windows-build]
+* MacOS amd64 [binary][mac-build]
+
+We can also provide you with a builds for:
```
Comment (svenklemm) [inline tutorials/telegraf-output-plugin.md]: remove `a`

File: tutorials/telegraf-output-plugin.md [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+ location         | text                     
+ ```
+
+ The `location` column was added and it contains "New York" for all rows.
+
+ ### Creating a separate metadata table for tags
+
+ The plugin we developed allows the user to select to have the tag sets inserted in a separate
+ table and then referenced via Foreign Keys in the measurement table. This way space can be saved
+ for measurements with low cardinality tag sets. To enable this change, you need to uncomment the `tags_as_foreignkeys` parameter in the plugin config (around line  103 in `telegraf.conf`) and set it to true
```
Comment (svenklemm) [inline tutorials/telegraf-output-plugin.md]: you also save space for high cardinality tags, and there are other benefits to having metadata in a separate table cause you can write certain queries more efficiently

File: tutorials/telegraf-output-plugin.md [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+Telegraf is written in Go, so only one standalone binary is required to run it. But also because of this all the code for the different plugins must be part of that binary. We have an unofficial build of Telegraf version 1.10.4 with our plugin added. You can download from the following links: 
+
+* Linux amd64: [deb][deb-build] [rpm][rpm-build] [binary][linux-bin-build]
+* Windows amd64 [binary/exe][windows-build]
+* MacOS amd64 [binary][mac-build]
+
+We can also provide you with a builds for:
+
+* Windows i386
+* Linux (i386, armhf, armel, arm64, static_amd64, s390x, mipsel
```
Comment (solugebefola) [inline tutorials/telegraf-output-plugin.md]: need closing parenthesis

File: tutorials/telegraf-output-plugin.md [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+* FreeBSD (amd64, i386)
+
+if you contact us using our [Public Support Slack][public-slack]
```
Comment (solugebefola) [inline tutorials/telegraf-output-plugin.md]: > contact us _in_ our [community slack][public-slack]

File: tutorials/telegraf-output-plugin.md [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+ table and then referenced via Foreign Keys in the measurement table. 
+ Having the tags as a separate table saves space for high cardinality tag sets, the insert rate is improved, and certain queries can be written to be more  efficeint. 
+ To enable this change, you need to uncomment the `tags_as_foreignkeys` parameter in the plugin config (around line  103 in `telegraf.conf`) and set it to true
+
+ ```
+ ## Store tags as foreign keys in the metrics table. Default is false.
+ tags_as_foreignkeys = true
+ ```
+
+ ### JSONB column for Tags and Fields
```
Comment (solugebefola) [inline tutorials/telegraf-output-plugin.md]: Can this occur at the same time as creating a separate metatable for tags?  If not, considering the user will be following along, you should mention switching `tags_as_foreignkeys` back to false.  If they can be used together, you should maybe make an example with tags in the separate table without JSONB in the section above and then have this example be for the combination of the two?

File: tutorials/telegraf-output-plugin.md [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+ cpu-total | 8.46385703620795
+ cpu0      | 12.4343351351033
+ cpu1      | 4.88380203380203
+ cpu2      | 12.2718724052057
+ cpu3      | 4.26716970050303
+```
+
+### Adding new Tags or Fields
+
+Your telegraf configuration can change at any moment. An input plugin can be reconfigured to produce different data, or you may decide to index your data with different tags. The plugin we built can dynamically update the created tables with new columns as they appear. The previous configuration we used had no global tags specified other than the `host` tag. We will now add a new global tag in the configuration. Open the file in any text editor and update the `[global_tags]` section (around line 18) with:
```
[the remark this answers is not in the document]
Comment (solugebefola) [reply tutorials/telegraf-output-plugin.md]: Also, please split sentences into separate lines for ease in reviewing (now and for future PRs)

File: tutorials/telegraf-output-plugin.md [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+For users of TimescaleDB they can update the `table_template` parameter in the config with
+```
+  table_template=`CREATE TABLE IF NOT EXISTS {TABLE}({COLUMNS}); SELECT create_hypertable({TABLELITERAL},'time',chunk_time_interval := '1 week'::interval,if_not_exists := true);`
+```
+
+This way when a new table is created it is converted into a hypertable, with each chunk holding 1 week intervals. Nothing else is needed to use the plugin with TimescaleDB. 
+
+## Running Telegraf
+
+When we run telegraf we only need to specify the config file to be used. If we execute
```
[the remark this answers is not in the document]
Comment (solugebefola) [reply tutorials/telegraf-output-plugin.md]: Capitalize "Telegraf"

File: tutorials/telegraf-output-plugin.md [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+```
+$  telegraf --config telegraf.conf 
+2019-05-23T13:48:09Z I! Starting Telegraf 1.10.4-with-pg
+2019-05-23T13:48:09Z I! Loaded inputs: cpu
+2019-05-23T13:48:09Z I! Loaded outputs: postgresql
+2019-05-23T13:48:09Z I! Tags enabled: host=local
+2019-05-23T13:48:09Z I! [agent] Config: Interval:10s, Quiet:false, Hostname:"local", Flush Interval:10s
+```
+
+In the output you can notice the loaded inputs (cpu) and outputs (postgresql) along with the global tags and the intervals with which the agent will collect the data from the inputs, and flush to the outputs. We can stop the execution of Telegraf after ~10-15 seconds.
```
[the remark this answers is not in the document]
Comment (solugebefola) [reply tutorials/telegraf-output-plugin.md]: `cpu`, `postgresql` in backticks

File: tutorials/telegraf-output-plugin.md [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+
+Let us now connect to our PostgreSQL instance and introspect the data
+
+```
+$ psql -U postgres -h localhost
+```
+
+The cpu input plugin has one measurement, called cpu, and it's stored in a table of the same name (by default in the public schema).
+So with the SQL query `SELECT * FROM cpu`, depending on how long you left Telegraf running you will see the table populated with some values, and we can find the average usage per cpu core with `SELECT cpu, avg(usage_user) FROM cpu GROUP BY cpu` and the output would look like
```
[the remark this answers is not in the document]
Comment (solugebefola) [reply tutorials/telegraf-output-plugin.md]: resolution here?

**Comment by solugebefola** [comment]
Some comments on phrasing were not addressed.  If you feel like the phrasing shouldn't change, please note that in a reply to the original comment

File: tutorials/telegraf-output-plugin.md [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+      1 | {"cpu": "cpu-total", "host": "local", "location": "New York"}
+      2 | {"cpu": "cpu0", "host": "local", "location": "New York"}
+      3 | {"cpu": "cpu1", "host": "local", "location": "New York"}
+```
+
+And instead of having three text columns, one JSONB column is created.
+
+## Next Steps
+
+Once you have started inserting data in TimescaleDB, you can begin to familiarize yourself with our [architecture](https://docs.timescale.com/introduction/architecture) and [API reference](https://docs.timescale.com/api).
```
Comment (solugebefola) [inline tutorials/telegraf-output-plugin.md]: Not critical, but these should be footnote links

File: tutorials/telegraf-output-plugin.md [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+ usage_softirq    | double precision         
+ usage_system     | double precision         
+ usage_iowait     | double precision         
+ usage_guest      | double precision         
+ usage_user       | double precision         
+ usage_idle       | double precision         
+ usage_steal      | double precision         
+
+```
+We can notice the `cpu`, `host` and `location` columns are not there, instead there's a `tag_id` column. The tag sets are stored in a separate table called `cpu_tag`:
```
Comment (solugebefola) [inline tutorials/telegraf-output-plugin.md]: >We can notice

"Notice that"

Not critical, I approved the PR.

File: tutorials/telegraf-output-plugin.md [hunk 14]
```diff
@@ -0,0 +1,346 @@
+# Collecting metrics with the PostgreSQL and TimescaleDB output plugin for Telegraf
+
+Telegraf can collect metrics from a wide array of inputs and write them to a wide array of outputs. 
+It is plugin-driven for both collection and output of data so it is easily extendable. 
+It is written in Go, which means that it is compiled and standalone binary that can be executed on any system with no need for external dependencies, or package management tools required. 
+
+Telegraf is an open-source tool. It contains over 200 plugins for gathering and writing different types of data written by people who work with that data. 
+
+We wrote the PostgreSQL output plugin which also has the ability to send data to an TimescaleDB hypertable. The [pull request][pull-request] is open and currently under review by the Telegraf developers, waiting to be merged. To give developers the opportunity to try this functionality, we built downloadable binaries of Telegraf with our plugin already included.
```
Comment (svenklemm) [inline tutorials/telegraf-output-plugin.md]: "an TimescaleDB hypertable" => "a TimescaleDB hypertable"

File: tutorials/telegraf-output-plugin.md [hunk 15]
```diff
@@ -0,0 +1,346 @@
...
+
+Once you download the binary and extract it to a suitable location (or install the packages) we can test out the build. 
+You may have to make the file executable by running `chmod +x telegraf`.
+We can check the version of the installed Telegraf with
+
+```
+$ telegraf --version
+```
+
+If the installation was good it should print out `Telegraf 1.10.4-with-pg`.
```
Comment (svenklemm) [inline tutorials/telegraf-output-plugin.md]: is good the right adjective here? i think successful seems more appropriate @solugebefola

File: tutorials/telegraf-output-plugin.md [hunk 16]
```diff
@@ -0,0 +1,346 @@
...
+ location         | text                     
+ ```
+
+ The `location` column was added and it contains "New York" for all rows.
+
+ ### Creating a separate metadata table for tags
+
+ The plugin we developed allows the user to select to have the tag sets inserted in a separate
+ table and then referenced via **foreign keys** in the measurement table. 
+ Having the tags in a separate table saves space for high cardinality tag sets, improves insert rate, and allows certain queries to be written more efficiently.
```
Comment (svenklemm) [inline tutorials/telegraf-output-plugin.md]: improves insert rate is wrong

File: tutorials/telegraf-output-plugin.md [hunk 17]
```diff
@@ -0,0 +1,346 @@
...
+ ## Use jsonb datatype for fields
+ fields_as_jsonb = false
+ ```
+
+To better visualize the result we'll drop the existing `cpu_tag` table from our database.
+```
+psql> DROP TABLE cpu_tag;
+```
+
+Fire up Telegraf up again, and turn it off after 20-30 seconds. Then we check the `cpu_tag` table:
```
Comment (svenklemm) [inline tutorials/telegraf-output-plugin.md]: 2x up, should probably be "Fire up Telegraf again"

File: tutorials/telegraf-output-plugin.md (hunk 15 above)
Comment (solugebefola) [reply tutorials/telegraf-output-plugin.md]: I like successful 👍

File: tutorials/telegraf-output-plugin.md [hunk 18]
```diff
@@ -344,3 +344,5 @@ Additionally, we have several other [tutorials][] available for you to explore a
 [rpm-build]: https://telegrafreleases.blob.core.windows.net/linux/telegraf-1.10.4~with~pg-1.x86_64.rpm
 [linux-bin-build]: https://telegrafreleases.blob.core.windows.net/linux/telegraf
 [pull-request]: https://github.com/influxdata/telegraf/pull/3428
+[architecture_lnk]: https://docs.timescale.com/introduction/architecture
```
Comment (solugebefola) [inline tutorials/telegraf-output-plugin.md]: These should be relative links `/introduction/architecture`, `/api`

**Pull request merged**